collector/lib/SysdigService.cpp

```diff
@@ -216,14 +232,23 @@ void SysdigService::Run(const std::atomic<ControlValue>& control) {
     if (!evt) continue;
 
     auto process_start = NowMicros();
-    for (auto& signal_handler : signal_handlers_) {
+    for (auto it = signal_handlers_.begin(); it != signal_handlers_.end(); it++) {
+      auto& signal_handler = *it;
       if (!signal_handler.ShouldHandle(evt)) continue;
       auto result = signal_handler.handler->HandleSignal(evt);
       if (result == SignalHandler::NEEDS_REFRESH) {
         if (!SendExistingProcesses(signal_handler.handler.get())) {
           continue;
         }
         result = signal_handler.handler->HandleSignal(evt);
+      } else if (result == SignalHandler::FINISHED) {
+        // This signal handler has finished processing events,
+        // so remove it from the signal handler list.
+        signal_handlers_.erase(it);
```

Comment:
If we continued to iterate, I think we'd have to save the next iterator but since we immediately break from the loop I think this is fine, though could do with some documentation describing what's going on here. I'll add something

Reply:
I see, thanks. It's probably worth adding some logging or metric around this. A consequence is that some handlers will not be called under the assumption it's a "self-check" event, but if something goes wrong it would be surprising to see no handlers invoked.